CWorld.cpp

```diff
@@ -3,11 +3,314 @@
+    for ( int i = 0; i < map.entities.size(); i++ ) {
+        const Entity& e = map.entities[ i ];
+        BaseGameEntity* baseEntity = NULL;
+        // Check the classname
+        for ( int j = 0; j < e.properties.size(); j++ ) {
+            const Property& prop = e.properties[ j ];
+            if ( prop.key == "classname" ) {
+                if ( prop.value == "func_door" ) {
+                    baseEntity = new Door(e.properties, e.brushes); 
+                    AddBrushesToDynamicGeometry( e.brushes );
+                    m_pEntityManager->RegisterEntity(baseEntity); 
+                    HKD_Model* model = ((Door*)baseEntity)->GetModel();
+                    m_BrushModels.push_back( model );
+                    std::vector<MapTri>& mapTris = ((Door*)baseEntity)->MapTris();
+                    m_pBrushMapTris.push_back( &mapTris );
+                } else if ( prop.value == "info_player_start" ) {
+                    assert( m_pPlayerEntity == nullptr ); // There can only be one
+                    glm::vec3 playerStartPosition = CWorld::GetOrigin(&e);
+                    m_pPlayerEntity = new Player(playerStartPosition);
+                    m_pEntityManager->RegisterEntity(m_pPlayerEntity);
+                    // Upload this model to the GPU. Not using the handle atm.
+                    int hPlayerModel = renderer->RegisterModel(m_pPlayerEntity->GetModel());
+                    m_Models.push_back(m_pPlayerEntity->GetModel());
+                } else if ( prop.value == "monster_soldier" ) {
+                    // just a placeholder entity from trenchbroom/quake
+                    glm::vec3 enemyStartPosition = CWorld::GetOrigin(&e);
+                    Enemy* enemy = new Enemy(e.properties);
+                    m_pEntityManager->RegisterEntity(enemy);
+
+                    int hEnemyModel = renderer->RegisterModel(enemy->GetModel());
+                    m_Models.push_back(enemy->GetModel());
+                } else if ( prop.value == "path_corner" ) { // FIX: Should be an entity type as well.
+                    Waypoint point = CWorld::GetWaypoint(&e);
+                    //m_pPath->AddPoint(point);
+                    // FIX: Very crude way of remembering waypoint names just so
+                    // that later on we can get its path.
+                    m_NameToWaypoint.insert({ point.targetname, point });
+                    // I assume that the corner Points are in the right order. if not we need to rethink the data structure
+                    glm::vec3 pathCornerPosition = CWorld::GetOrigin(&e);
+                    printf("Path corner entity found: %f, %f, %f\n",
+                           pathCornerPosition.x,
+                           pathCornerPosition.y,
+                           pathCornerPosition.z);
+                } else {
+                    printf("Unknown entity type: %s\n", prop.value.c_str());
+                }
+            }
+        }
+    }
```

Comment:
Not yet. Not all entities are in place, I don't know the requirements yet for semantically compressing this. The area under the ifs is not large enough to be unreadable either. 
See: http://number-none.com/blow/blog/programming/2014/09/26/carmack-on-inlined-code.html

Reply:
oh thanks, very good read.

i see where this can become problematic. offtopic: do you see this as a c/cpp problem where backtraces and mutability are so easy and produce impossible state quickly or is this a general problem?

what do you think of unit tests in cases like this? in a client project where we are not as concerned of performance as we are when writing realtime graphics stuff. i would have written a unit test or similar to make sure the function does what it should do. (while writing this, it came to me that it also only makes sense for pure functions)